app/src/main/java/chat/rocket/android/wallet/create/presentation/CreateWalletPresenter.kt

```diff
@@ -0,0 +1,89 @@
+                                                 serverInteractor: GetCurrentServerInteractor,
+                                                 factory: RocketChatClientFactory){
+
+    private val serverUrl = serverInteractor.get()!!
+    private val client: RocketChatClient = factory.create(serverUrl)
+    private val restUrl: HttpUrl? = HttpUrl.parse(serverUrl)
+    private val bcInterface = BlockchainInterface()
+
+    /**
+     * Create a bip39 wallet (which uses mnemonic passphrases)
+     *  and save the address to the user's account and display the passphrases to the user
+     */
+    fun createNewWallet(walletName: String, password: String){
+        launchUI(strategy) {
+            try {
+                val response = bcInterface.createBip39Wallet(password, view.returnContext())
+                val address = response[0]
+                val mnemonic =  response[1]
+                updateWalletAddress(address, {
+                    view.showWalletSuccessfullyCreatedMessage(mnemonic)
+                })
+            } catch (exception: Exception) {
+                view.showWalletCreationFailedMessage(exception.message)
+            }
+        }
+    }
+
+    /**
+     * Change the walletAddress field in the current user's customFields field
+     *
+     * @param address user's new wallet address
+     *
+     * NOTE: this function directly calls the REST API, which normally should be
+     *          done in the Kotlin SDK
+     */
+    private fun updateWalletAddress(address: String, callback: () -> Unit) {
+        launchUI(strategy) {
+            try {
+                val httpUrl = restUrl?.newBuilder()
+                        ?.addPathSegment("api")
+                        ?.addPathSegment("v1")
+                        ?.addPathSegment("users.update")
+                        ?.build()
+
+                val me = retryIO("me") { client.me() }
+                val payloadBody = "{\"userId\":\"${me.id}\",\"data\":{\"customFields\":{\"walletAddress\":\"$address\"}}}"
+                val body = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), payloadBody)
+                val builder = Request.Builder().url(httpUrl)
+
+                val token: Token? = tokenRepository.get(serverUrl)
+                token?.let {
+                    builder.addHeader("X-Auth-Token", token.authToken)
+                            .addHeader("X-User-Id", token.userId)
+                }
+                val request = builder.post(body).build()
+
+                val httpClient = OkHttpClient()
+                httpClient.newCall(request).enqueue(object : Callback {
+                    override fun onFailure(call: Call, e: IOException) { Timber.d("ERROR: request call failed!")}
+                    override fun onResponse(call: Call, response: Response) {
+                        callback()
+                    }
+                })
+            } catch (ex: Exception) {
+                Timber.e(ex)
```

Comment:
It seems like if something goes wrong in this function, that we will just hang the ui on the screen it was on when the create wallet button was clicked, because we catch exceptions and don't call any failure callback.